File: browser/ai_browser_agent/lib/agent_server.mjs
import fs from 'node:fs/promises';
import { readFileSync } from 'node:fs';
import http from 'node:http';
import path from 'node:path';
import { randomBytes } from 'node:crypto';
import { fileURLToPath } from 'node:url';

import {
  AiBrowserAgentConfig,
  AiBrowserState,
  AiRuntimeState,
  AiToolExecutionContext,
  AiToolRegistry,
  clampNumber,
  ensureJsonObject,
} from './models.mjs';
import { ResultHandleStore } from './result_handle_store.mjs';
import { defaultBundles, registerDefaultTools } from './tools.mjs';

export class AiBrowserAgentServer {
  constructor(config) {
    this.config = config;
    this.runtimeState = new AiRuntimeState();
    this.resultHandleStore = new ResultHandleStore();
    this.browserState = new AiBrowserState();
    this.registry = new AiToolRegistry({ bundles: defaultBundles() });
    this.server = null;
    this.initialized = false;
    this.token = '';
    this.agentManual = readFileSync(
      path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..', 'AGENT.md'),
      'utf8',
    );
    this.toolContext = new AiToolExecutionContext({
      config: this.config,
      registry: this.registry,
      resultHandleStore: this.resultHandleStore,
      runtimeState: this.runtimeState,
      browserState: this.browserState,
      agentManual: this.agentManual,
      readToken: () => this.token,
      regenerateToken: async () => this.#regenerateToken(),
      buildHealthPayload: ({ includeOk = false } = {}) => this.#buildHealthPayload({ includeOk }),
      buildAgentBriefPayload: ({ includeOk = false } = {}) => this.#buildAgentBriefPayload({ includeOk }),
    });
  }

  get isRunning() {
    return this.server !== null;
  }

  async start() {
    if (this.isRunning) {
      return;
    }
    if (!this.initialized) {
      registerDefaultTools(this.registry, this.toolContext);
      this.initialized = true;
    }
    if (this.config.requireToken) {
      this.token = await this.#ensureToken();
    }
    this.server = http.createServer((request, response) => {
      this.#handleRequest(request, response).catch(async (error) => {
        this.runtimeState.log('error', String(error));
        await this.#sendError(response, 500, String(error));
      });
    });
    await new Promise((resolve, reject) => {
      this.server.on('error', reject);
      this.server.listen(this.config.port, this.config.host, resolve);
    });
    this.runtimeState.log('info', `Service started at ${this.config.serverUrl}`);
  }

  async stop() {
    if (!this.server) {
      return;
    }
    const server = this.server;
    this.server = null;
    await new Promise((resolve, reject) => {
      server.close((error) => {
        if (error) {
          reject(error);
        } else {
          resolve();
        }
      });
    });
  }

  async #handleRequest(request, response) {
    this.#addCommonHeaders(response);
    if (request.method === 'OPTIONS') {
      response.writeHead(204);
      response.end();
      return;
    }

    const parsedUrl = new URL(request.url, this.config.serverUrl);
    const pathName = parsedUrl.pathname.replace(/^\/+/, '');

    if (pathName === 'health' && request.method === 'GET') {
      await this.#sendJson(response, 200, this.#buildHealthPayload({ includeOk: true }));
      return;
    }

    if (!this.#isAuthorized(request)) {
      await this.#sendError(
        response,
        401,
        `Unauthorized. Provide ${AiBrowserAgentConfig.primaryTokenHeader} or ${AiBrowserAgentConfig.legacyTokenHeader}.`,
      );
      return;
    }

    if ((pathName === 'manifest' || pathName === 'manifest/summary') && request.method === 'GET') {
      const detail = parsedUrl.searchParams.get('detail') ?? '';
      const payload = detail.toLowerCase() === 'full'
        ? this.registry.buildManifestFull(this.config)
        : this.registry.buildManifestSummary(this.config);
      await this.#sendJson(response, 200, payload);
      return;
    }

    if (pathName === 'manifest/full' && request.method === 'GET') {
      await this.#sendJson(response, 200, this.registry.buildManifestFull(this.config));
      return;
    }

    if (pathName === 'manifest/bundles' && request.method === 'GET') {
      await this.#sendJson(response, 200, this.registry.buildBundleIndex(this.config));
      return;
    }

    if (pathName.startsWith('manifest/bundle/') && request.method === 'GET') {
      const bundleId = decodeURIComponent(pathName.slice('manifest/bundle/'.length));
      const payload = this.registry.tryBuildBundle(this.config, bundleId);
      if (!payload) {
        await this.#sendError(response, 404, `Unknown manifest bundle: ${bundleId}`);
        return;
      }
      await this.#sendJson(response, 200, payload);
      return;
    }

    if (pathName === 'manifest/search' && request.method === 'POST') {
      const body = await this.#readBodyAsJson(request);
      const payload = this.registry.buildManifestSearch(this.config, {
        query: body.query ? String(body.query) : '',
        limit: body.limit,
        namespaceId: body.namespaceId ? String(body.namespaceId) : '',
        bundleId: body.bundleId ? String(body.bundleId) : '',
      });
      await this.#sendJson(response, 200, payload);
      return;
    }

    if (pathName === 'tool/describe_many' && request.method === 'POST') {
      const body = await this.#readBodyAsJson(request);
      await this.#sendJson(
        response,
        200,
        this.registry.buildDescribeMany(this.config, Array.isArray(body.ids) ? body.ids : []),
      );
      return;
    }

    if (pathName === 'agent/brief' && request.method === 'GET') {
      await this.#sendJson(response, 200, this.#buildAgentBriefPayload({ includeOk: true }));
      return;
    }

    if (pathName === 'agent' && request.method === 'GET') {
      await this.#sendJson(response, 200, {
        ok: true,
        content: this.agentManual,
      });
      return;
    }

    if (pathName.startsWith('result/') && request.method === 'GET') {
      const handleId = decodeURIComponent(pathName.slice('result/'.length));
      const payload = this.resultHandleStore.buildPage(handleId, {
        offset: clampNumber(Number(parsedUrl.searchParams.get('offset')) || 0, 0, Number.MAX_SAFE_INTEGER),
        limit: clampNumber(Number(parsedUrl.searchParams.get('limit')) || 0, 0, 32768),
      });
      if (!payload) {
        await this.#sendError(response, 404, `Unknown result handle: ${handleId}`);
        return;
      }
      await this.#sendJson(response, 200, payload);
      return;
    }

    if (pathName.startsWith('call/') && request.method === 'POST') {
      const toolId = decodeURIComponent(pathName.slice('call/'.length));
      const tool = this.registry.findTool(toolId);
      if (!tool) {
        await this.#sendJson(response, 404, {
          ok: false,
          toolId,
          error: `Unknown tool: ${toolId}`,
        });
        return;
      }
      if (tool.requiresConfirmation && !this.config.fullAccessEnabled) {
        await this.#sendJson(response, 403, {
          ok: false,
          toolId,
          error: 'High-risk tool requires --full-access for this adapter.',
        });
        return;
      }
      const body = await this.#readBodyAsJson(request);
      const startedAt = Date.now();
      try {
        const result = await Promise.race([
          tool.handler(body, this.toolContext),
          timeoutAfter(this.config.toolTimeoutMs),
        ]);
        this.runtimeState.recordCall({
          toolId,
          ok: true,
          durationMs: Date.now() - startedAt,
          message: 'ok',
        });
        await this.#sendJson(response, 200, {
          ok: true,
          toolId,
          durationMs: Date.now() - startedAt,
          result,
        });
      } catch (error) {
        this.runtimeState.recordCall({
          toolId,
          ok: false,
          durationMs: Date.now() - startedAt,
          message: String(error),
        });
        await this.#sendJson(response, 500, {
          ok: false,
          toolId,
          durationMs: Date.now() - startedAt,
          error: String(error),
        });
      }
      return;
    }

    await this.#sendError(response, 404, `Not found: ${pathName}`);
  }

  #isAuthorized(request) {
    if (!this.config.requireToken) {
      return true;
    }
    const provided = request.headers[AiBrowserAgentConfig.primaryTokenHeader.toLowerCase()]
      ?? request.headers[AiBrowserAgentConfig.legacyTokenHeader.toLowerCase()];
    return typeof provided === 'string' && provided === this.token;
  }

  async #readBodyAsJson(request) {
    const chunks = [];
    for await (const chunk of request) {
      chunks.push(chunk);
    }
    const content = Buffer.concat(chunks).toString('utf8');
    if (!content.trim()) {
      return {};
    }
    return ensureJsonObject(JSON.parse(content));
  }

  #buildHealthPayload({ includeOk = false } = {}) {
    return {
      ...(includeOk ? { ok: true } : {}),
      framework: AiBrowserAgentConfig.frameworkName,
      service: AiBrowserAgentConfig.serviceName,
      serviceId: AiBrowserAgentConfig.serviceId,
      version: this.config.serviceVersion,
      platformId: AiBrowserAgentConfig.platformId,
      protocolVersion: AiBrowserAgentConfig.protocolVersion,
      serverRunning: this.isRunning,
      requiresToken: this.config.requireToken,
      acceptedTokenHeaders: this.config.acceptedTokenHeaders,
      serverUrl: this.config.serverUrl,
      manifestHash: this.registry.manifestHash,
      toolCount: this.registry.count,
      namespaces: this.registry.namespaceInfos,
      supportsManifestSearch: true,
      supportsDescribeMany: true,
      supportsResultHandles: true,
      supportsBundles: true,
      supportsTextChunking: true,
      supportsDynamicToolRegistration: false,
      recommendedFlow: [
        'GET /health and compare manifestHash before refreshing capabilities.',
        'Use POST /manifest/search or GET /manifest/bundle/{id} to narrow candidate tools.',
        'Use POST /tool/describe_many for exact argument and return schemas before calling tools.',
        'Use GET /result/{handleId} for additional pages or text chunks when a tool returns a resultHandle.',
        'Use GET /manifest/full only as a fallback when search is insufficient.',
      ],
      paths: {
        health: '/health',
        manifestSummary: '/manifest',
        manifestFull: '/manifest/full',
        manifestSearch: '/manifest/search',
        manifestBundles: '/manifest/bundles',
        toolDescribeMany: '/tool/describe_many',
        call: '/call/{toolId}',
        agent: '/agent',
        agentBrief: '/agent/brief',
        resultPage: '/result/{handleId}',
      },
      platform: {
        chromeBaseUrl: this.config.chromeBaseUrl,
        currentTargetId: this.browserState.currentTargetId || null,
        lastLaunchInfo: this.browserState.lastLaunchInfo,
      },
    };
  }

  #buildAgentBriefPayload({ includeOk = false } = {}) {
    return {
      ...(includeOk ? { ok: true } : {}),
      framework: AiBrowserAgentConfig.frameworkName,
      platformId: AiBrowserAgentConfig.platformId,
      summary: 'Prefer cached discovery through health, manifest search, on-demand tool descriptions, and paged result handles instead of repeatedly loading the full manifest.',
      steps: [
        'Call GET /health and reuse cached capabilities while manifestHash stays unchanged.',
        'Search tools with POST /manifest/search or load a focused bundle with GET /manifest/bundle/{id}.',
        'Request exact tool schemas with POST /tool/describe_many before calling unfamiliar tools.',
        'When a tool returns resultHandle, page additional data through GET /result/{handleId} instead of re-running the tool with larger limits.',
        'Fallback to GET /manifest/full or GET /agent only when the optimized discovery flow is insufficient.',
      ],
      paths: {
        health: '/health',
        manifestSummary: '/manifest',
        manifestFull: '/manifest/full',
        manifestSearch: '/manifest/search',
        manifestBundles: '/manifest/bundles',
        toolDescribeMany: '/tool/describe_many',
        call: '/call/{toolId}',
        agent: '/agent',
        agentBrief: '/agent/brief',
        resultPage: '/result/{handleId}',
      },
    };
  }

  async #ensureToken() {
    await fs.mkdir(path.dirname(this.config.tokenFilePath), { recursive: true });
    try {
      const existing = (await fs.readFile(this.config.tokenFilePath, 'utf8')).trim();
      if (existing) {
        return existing;
      }
    } catch {
      // create below
    }
    return this.#regenerateToken();
  }

  async #regenerateToken() {
    const token = randomBytes(24).toString('hex');
    await fs.mkdir(path.dirname(this.config.tokenFilePath), { recursive: true });
    await fs.writeFile(this.config.tokenFilePath, token, 'utf8');
    this.token = token;
    this.runtimeState.log('info', 'Token regenerated.');
    return token;
  }

  #addCommonHeaders(response) {
    response.setHeader('Access-Control-Allow-Origin', '*');
    response.setHeader('Access-Control-Allow-Headers', 'Content-Type, X-AI-Agent-Token, X-Browser-Ai-Token');
    response.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
    response.setHeader('Content-Type', 'application/json; charset=utf-8');
  }

  async #sendJson(response, statusCode, payload) {
    response.writeHead(statusCode);
    response.end(JSON.stringify(payload, null, 2));
  }

  async #sendError(response, statusCode, message) {
    await this.#sendJson(response, statusCode, {
      ok: false,
      error: message,
    });
  }
}

function timeoutAfter(timeoutMs) {
  return new Promise((_, reject) => {
    setTimeout(() => reject(new Error(`Tool timed out after ${timeoutMs}ms`)), timeoutMs);
  });
}

File: flutter/ai_flutter_agent/lib/src/agent_server.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';

import 'generated_tool_host.dart';
import 'models.dart';
import 'result_handle_store.dart';
import 'tools.dart';

class AiFlutterAgentServer {
  AiFlutterAgentServer(this.config)
      : runtimeState = AiRuntimeState(),
        resultHandleStore = ResultHandleStore(),
        registry = AiToolRegistry(bundles: defaultBundles()),
        generatedToolHost = AiFlutterGeneratedToolHost(
          config: config,
        );

  final AiFlutterAgentConfig config;
  final AiRuntimeState runtimeState;
  final ResultHandleStore resultHandleStore;
  final AiToolRegistry registry;
  final AiFlutterGeneratedToolHost generatedToolHost;

  HttpServer? _server;
  StreamSubscription<FileSystemEvent>? _generatedToolsWatcher;
  bool _generatedToolsDirty = true;
  String _generatedToolsFingerprint = '';
  int _generatedToolsCount = 0;
  late final AiToolExecutionContext _toolContext = AiToolExecutionContext(
    config: config,
    registry: registry,
    resultHandleStore: resultHandleStore,
    runtimeState: runtimeState,
    agentManual: _agentManual,
    readToken: () => _token,
    regenerateToken: _regenerateToken,
    buildHealthPayload: ({bool includeOk = false}) => _buildHealthPayload(includeOk: includeOk),
    buildAgentBriefPayload: ({bool includeOk = false}) => _buildAgentBriefPayload(includeOk: includeOk),
    generatedToolHost: generatedToolHost,
    reloadGeneratedTools: ({bool force = false}) => _refreshRegistry(force: force),
  );
  late final String _agentManual = _loadAgentManual();
  String _token = '';

  bool get isRunning => _server != null;

  Future<void> start() async {
    if (isRunning) {
      return;
    }
    await generatedToolHost.ensureLayout();
    await _refreshRegistry(force: true);
    if (config.requireToken) {
      _token = await _ensureToken();
    }
    _server = await HttpServer.bind(config.host, config.port);
    _generatedToolsWatcher = Directory(config.generatedToolsDirectoryPath)
        .watch()
        .listen(
          (event) => markGeneratedToolsDirty('file event: ${event.type}'),
          onError: (Object error) => runtimeState.log(
            'warning',
            'Generated tool watcher error: $error',
          ),
        );
    runtimeState.log('info', 'Service started at ${config.serverUrl}');
    unawaited(_server!.forEach(_handleRequest));
  }

  Future<void> stop() async {
    if (_server == null) {
      return;
    }
    runtimeState.log('info', 'Service stopping.');
    await _generatedToolsWatcher?.cancel();
    _generatedToolsWatcher = null;
    await _server!.close(force: true);
    _server = null;
  }

  void markGeneratedToolsDirty([String reason = '']) {
    _generatedToolsDirty = true;
    if (reason.trim().isNotEmpty) {
      runtimeState.log('info', 'Generated tool registry marked dirty: $reason');
    }
  }

  Future<void> _handleRequest(HttpRequest request) async {
    _addCommonHeaders(request.response);

    if (request.method == 'OPTIONS') {
      request.response
        ..statusCode = HttpStatus.noContent
        ..close();
      return;
    }

    try {
      await _refreshRegistry();
      final path = request.uri.pathSegments.join('/');
      if (path == 'health' && request.method == 'GET') {
        await _sendJson(request.response, HttpStatus.ok, _buildHealthPayload(includeOk: true));
        return;
      }

      if (!_isAuthorized(request)) {
        await _sendError(
          request.response,
          HttpStatus.unauthorized,
          'Unauthorized. Provide ${AiFlutterAgentConfig.primaryTokenHeader} or ${AiFlutterAgentConfig.legacyTokenHeader}.',
        );
        return;
      }

      if ((path == 'manifest' || path == 'manifest/summary') && request.method == 'GET') {
        final detail = request.uri.queryParameters['detail'] ?? '';
        final payload = detail.toLowerCase() == 'full'
            ? registry.buildManifestFull(config)
            : registry.buildManifestSummary(config);
        await _sendJson(request.response, HttpStatus.ok, payload);
        return;
      }

      if (path == 'manifest/full' && request.method == 'GET') {
        await _sendJson(request.response, HttpStatus.ok, registry.buildManifestFull(config));
        return;
      }

      if (path == 'manifest/bundles' && request.method == 'GET') {
        await _sendJson(request.response, HttpStatus.ok, registry.buildBundleIndex(config));
        return;
      }

      if (path.startsWith('manifest/bundle/') && request.method == 'GET') {
        final bundleId = Uri.decodeComponent(path.substring('manifest/bundle/'.length));
        final payload = registry.tryBuildBundle(config, bundleId);
        if (payload == null) {
          await _sendError(request.response, HttpStatus.notFound, 'Unknown manifest bundle: $bundleId');
          return;
        }
        await _sendJson(request.response, HttpStatus.ok, payload);
        return;
      }

      if (path == 'manifest/search' && request.method == 'POST') {
        final body = await _readBodyAsJson(request);
        final payload = registry.buildManifestSearch(
          config,
          query: body['query']?.toString() ?? '',
          limit: _parseInt(body['limit'], 0),
          namespaceId: body['namespaceId']?.toString() ?? '',
          bundleId: body['bundleId']?.toString() ?? '',
        );
        await _sendJson(request.response, HttpStatus.ok, payload);
        return;
      }

      if (path == 'tool/describe_many' && request.method == 'POST') {
        final body = await _readBodyAsJson(request);
        final ids = (body['ids'] as List?)?.map((item) => item.toString());
        await _sendJson(request.response, HttpStatus.ok, registry.buildDescribeMany(config, ids));
        return;
      }

      if (path == 'agent/brief' && request.method == 'GET') {
        await _sendJson(request.response, HttpStatus.ok, _buildAgentBriefPayload(includeOk: true));
        return;
      }

      if (path == 'agent' && request.method == 'GET') {
        await _sendJson(request.response, HttpStatus.ok, <String, dynamic>{
          'ok': true,
          'content': _agentManual,
        });
        return;
      }

      if (path.startsWith('result/') && request.method == 'GET') {
        final handleId = Uri.decodeComponent(path.substring('result/'.length));
        final payload = resultHandleStore.buildPage(
          handleId,
          offset: _parseInt(request.uri.queryParameters['offset'], 0),
          limit: _parseInt(request.uri.queryParameters['limit'], 0),
        );
        if (payload == null) {
          await _sendError(request.response, HttpStatus.notFound, 'Unknown result handle: $handleId');
          return;
        }
        await _sendJson(request.response, HttpStatus.ok, payload);
        return;
      }

      if (path.startsWith('call/') && request.method == 'POST') {
        final toolId = Uri.decodeComponent(path.substring('call/'.length));
        final body = await _readBodyAsJson(request);
        final tool = registry.findTool(toolId);
        if (tool == null) {
          await _sendJson(
            request.response,
            HttpStatus.notFound,
            <String, dynamic>{
              'ok': false,
              'toolId': toolId,
              'error': 'Unknown tool: $toolId',
            },
          );
          return;
        }

        if (tool.requiresConfirmation && !config.fullAccessEnabled) {
          await _sendJson(
            request.response,
            HttpStatus.forbidden,
            <String, dynamic>{
              'ok': false,
              'toolId': toolId,
              'error': 'High-risk tool requires --full-access for this adapter.',
            },
          );
          return;
        }

        final watch = Stopwatch()..start();
        try {
          final result = await tool.handler(body, _toolContext)
              .timeout(Duration(milliseconds: config.toolTimeoutMs));
          runtimeState.recordCall(
            toolId: toolId,
            ok: true,
            durationMs: watch.elapsedMilliseconds,
            message: 'ok',
          );
          await _sendJson(
            request.response,
            HttpStatus.ok,
            <String, dynamic>{
              'ok': true,
              'toolId': toolId,
              'durationMs': watch.elapsedMilliseconds,
              'result': result,
            },
          );
        } catch (error) {
          runtimeState.recordCall(
            toolId: toolId,
            ok: false,
            durationMs: watch.elapsedMilliseconds,
            message: error.toString(),
          );
          await _sendJson(
            request.response,
            HttpStatus.internalServerError,
            <String, dynamic>{
              'ok': false,
              'toolId': toolId,
              'durationMs': watch.elapsedMilliseconds,
              'error': error.toString(),
            },
          );
        }
        return;
      }

      await _sendError(request.response, HttpStatus.notFound, 'Not found: $path');
    } catch (error) {
      runtimeState.log('error', error.toString());
      await _sendError(request.response, HttpStatus.internalServerError, error.toString());
    }
  }

  bool _isAuthorized(HttpRequest request) {
    if (!config.requireToken) {
      return true;
    }
    final provided = request.headers.value(AiFlutterAgentConfig.primaryTokenHeader) ??
        request.headers.value(AiFlutterAgentConfig.legacyTokenHeader);
    return provided != null && provided == _token;
  }

  Future<JsonMap> _readBodyAsJson(HttpRequest request) async {
    final content = await utf8.decoder.bind(request).join();
    if (content.trim().isEmpty) {
      return <String, dynamic>{};
    }
    final decoded = jsonDecode(content);
    return ensureJsonMap(decoded);
  }

  JsonMap _buildHealthPayload({bool includeOk = false}) {
    return <String, dynamic>{
      if (includeOk) 'ok': true,
      'framework': AiFlutterAgentConfig.frameworkName,
      'service': AiFlutterAgentConfig.serviceName,
      'serviceId': AiFlutterAgentConfig.serviceId,
      'version': config.serviceVersion,
      'platformId': AiFlutterAgentConfig.platformId,
      'protocolVersion': AiFlutterAgentConfig.protocolVersion,
      'serverRunning': isRunning,
      'requiresToken': config.requireToken,
      'acceptedTokenHeaders': config.acceptedTokenHeaders,
      'serverUrl': config.serverUrl,
      'manifestHash': registry.manifestHash,
      'toolCount': registry.count,
      'namespaces': registry.namespaceInfos,
      'supportsManifestSearch': true,
      'supportsDescribeMany': true,
      'supportsResultHandles': true,
      'supportsBundles': true,
      'supportsTextChunking': true,
      'supportsDynamicToolRegistration': true,
      'recommendedFlow': <String>[
        'GET /health and compare manifestHash before refreshing capabilities.',
        'Use POST /manifest/search or GET /manifest/bundle/{id} to narrow candidate tools.',
        'Use POST /tool/describe_many for exact argument and return schemas before calling tools.',
        'Use GET /result/{handleId} for additional pages or text chunks when a tool returns a resultHandle.',
        'Use GET /manifest/full only as a fallback when search is insufficient.',
      ],
      'paths': <String, dynamic>{
        'health': '/health',
        'manifestSummary': '/manifest',
        'manifestFull': '/manifest/full',
        'manifestSearch': '/manifest/search',
        'manifestBundles': '/manifest/bundles',
        'toolDescribeMany': '/tool/describe_many',
        'call': '/call/{toolId}',
        'agent': '/agent',
        'agentBrief': '/agent/brief',
        'resultPage': '/result/{handleId}',
      },
      'platform': <String, dynamic>{
        'projectRoot': config.projectRoot,
        'flutterExecutable': config.flutterExecutable,
        'dartExecutable': config.dartExecutable,
        'generatedToolsDirectoryPath': config.generatedToolsDirectoryPath,
        'generatedToolRuntimeDirectoryPath': config.generatedToolRuntimeDirectoryPath,
        'generatedToolCount': _generatedToolsCount,
      },
    };
  }

  JsonMap _buildAgentBriefPayload({bool includeOk = false}) {
    return <String, dynamic>{
      if (includeOk) 'ok': true,
      'framework': AiFlutterAgentConfig.frameworkName,
      'platformId': AiFlutterAgentConfig.platformId,
      'summary': 'Prefer cached discovery through health, manifest search, on-demand tool descriptions, and paged result handles instead of repeatedly loading the full manifest.',
      'steps': <String>[
        'Call GET /health and reuse cached capabilities while manifestHash stays unchanged.',
        'Search tools with POST /manifest/search or load a focused bundle with GET /manifest/bundle/{id}.',
        'Request exact tool schemas with POST /tool/describe_many before calling unfamiliar tools.',
        'When a tool returns resultHandle, page additional data through GET /result/{handleId} instead of re-running the tool with larger limits.',
        'Fallback to GET /manifest/full or GET /agent only when the optimized discovery flow is insufficient.',
      ],
      'paths': <String, dynamic>{
        'health': '/health',
        'manifestSummary': '/manifest',
        'manifestFull': '/manifest/full',
        'manifestSearch': '/manifest/search',
        'manifestBundles': '/manifest/bundles',
        'toolDescribeMany': '/tool/describe_many',
        'call': '/call/{toolId}',
        'agent': '/agent',
        'agentBrief': '/agent/brief',
        'resultPage': '/result/{handleId}',
      },
    };
  }

  Future<String> _ensureToken() async {
    final tokenFile = File(config.tokenFilePath);
    await tokenFile.parent.create(recursive: true);
    if (await tokenFile.exists()) {
      final existing = (await tokenFile.readAsString()).trim();
      if (existing.isNotEmpty) {
        return existing;
      }
    }
    return _regenerateToken();
  }

  Future<String> _regenerateToken() async {
    final token = _generateToken();
    final tokenFile = File(config.tokenFilePath);
    await tokenFile.parent.create(recursive: true);
    await tokenFile.writeAsString(token);
    _token = token;
    runtimeState.log('info', 'Token regenerated.');
    return token;
  }

  String _generateToken() {
    final random = Random.secure();
    final buffer = StringBuffer();
    for (var index = 0; index < 32; index++) {
      buffer.write(random.nextInt(16).toRadixString(16));
    }
    return buffer.toString();
  }

  String _loadAgentManual() {
    try {
      final manualUri = Platform.script.resolve('../AGENT.md');
      final file = File.fromUri(manualUri);
      if (file.existsSync()) {
        return file.readAsStringSync();
      }
    } catch (_) {
    }
    return '''
# AI Flutter Agent Protocol

Use GET /health, cache manifestHash, prefer manifest search or bundle loading, then use tool/describe_many before calling unfamiliar tools.
''';
  }

  void _addCommonHeaders(HttpResponse response) {
    response.headers
      ..set('Access-Control-Allow-Origin', 'http://127.0.0.1')
      ..set('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
      ..set(
        'Access-Control-Allow-Headers',
        'Content-Type, ${AiFlutterAgentConfig.primaryTokenHeader}, ${AiFlutterAgentConfig.legacyTokenHeader}',
      )
      ..set('Cache-Control', 'no-store')
      ..contentType = ContentType.json;
  }

  Future<void> _sendError(HttpResponse response, int statusCode, String message) {
    return _sendJson(response, statusCode, <String, dynamic>{'ok': false, 'error': message});
  }

  Future<void> _sendJson(HttpResponse response, int statusCode, Object? payload) async {
    response.statusCode = statusCode;
    response.write(jsonEncode(payload));
    await response.close();
  }

  Future<JsonMap> _refreshRegistry({bool force = false}) async {
    if (!force && !_generatedToolsDirty && registry.count > 0) {
      return <String, dynamic>{
        'reloaded': false,
        'manifestHash': registry.manifestHash,
        'generatedToolCount': _generatedToolsCount,
      };
    }

    final fingerprint = await generatedToolHost.computeFingerprint();
    if (!force && fingerprint == _generatedToolsFingerprint && registry.count > 0) {
      _generatedToolsDirty = false;
      return <String, dynamic>{
        'reloaded': false,
        'manifestHash': registry.manifestHash,
        'generatedToolCount': _generatedToolsCount,
      };
    }

    registry.reset();
    _toolContext.registry = registry;
    registerDefaultTools(registry, _toolContext);
    final generatedDefinitions = await generatedToolHost.loadDefinitions();
    for (final definition in generatedDefinitions) {
      registry.register(generatedToolHost.createToolDefinition(definition, _toolContext));
    }
    _generatedToolsFingerprint = fingerprint;
    _generatedToolsDirty = false;
    _generatedToolsCount = generatedDefinitions.length;
    runtimeState.log(
      'info',
      'Generated tool registry refreshed. Loaded ${generatedDefinitions.length} generated tool(s).',
    );
    return <String, dynamic>{
      'reloaded': true,
      'manifestHash': registry.manifestHash,
      'generatedToolCount': generatedDefinitions.length,
    };
  }
}

int _parseInt(Object? value, int defaultValue) {
  if (value is int) {
    return value;
  }
  if (value is num) {
    return value.toInt();
  }
  if (value is String) {
    return int.tryParse(value) ?? defaultValue;
  }
  return defaultValue;
}
